sidebar/panel.js

```diff
@@ -526,37 +546,66 @@ function start() {
     },
 
     getCurrentSpeed(item) {
+      const { PAUSED, COMPLETE } = this.states;
       const downloadState = this.calculateDownloadState(item);
-      if (downloadState === PAUSED) {
+      if (downloadState === PAUSED || downloadState === COMPLETE) {
         return "0";
       }
+
       const remainingSeconds = this.getRemainingSeconds(item);
-      if (remainingSeconds === 0) {
-        return "0";
+
+      if (isNaN(remainingSeconds)){
+        return "Calculating"
+      } else if (remainingSeconds === 0) {
+        return "Finishing";
+      }
+
+      const remainingBytes = item.totalBytes - item.bytesReceived;
+      const currentSpeed = Math.round(remainingBytes / remainingSeconds);
+
+      if (currentSpeed > SPEED_GIGABYTE) {
+        return this.divideAndRound(currentSpeed, SPEED_GIGABYTE) + " GB/s";
+      } else if (currentSpeed > SPEED_MEGABYTE) {
+        return this.divideAndRound(currentSpeed, SPEED_MEGABYTE) + " MB/s";
       }
 
-      const remainingMegaBytes = (item.totalBytes - item.bytesReceived) / 1048576;
-      const currentSpeed = Math.round(remainingMegaBytes / remainingSeconds);
-      return this.checkMinuteToString(currentSpeed);
+      const kiloSpeed = this.divideAndRound(currentSpeed, SPEED_KILOBYTE);
+
+      if (kiloSpeed <= 0) {
+        return "< 1 KB/s";
+      }
+
+      return kiloSpeed + " KB/s";
     },
 
-    getRemainingMinutesString(item) {
+    getRemainingTimeString(item) {
+      const { PAUSED, COMPLETE } = this.states;
       const downloadState = this.calculateDownloadState(item);
       if (downloadState === PAUSED) {
         return "Paused";
+      } else if (downloadState === COMPLETE) {
+        return "Completed";
       }
-      const differenceMinutes = this.getRemainingMinutes(item);
-      const remainingMinutes = this.checkMinuteToString(differenceMinutes);
-      
-      if (remainingMinutes === "<0"){
-        return "Less than a minute remaining";
+
+      const remainingSeconds = this.getRemainingSeconds(item);
+      const prefixSeparator = "- ";
+
+      //this is quite ad-hoc, but it is the best possible solution
```

Comment:
You could make it a bit more readable this way, what do you think?
```javascript
if (remainingSeconds > DAY_INT) {
  return "Over a day remaining";
}

let timeUnit = 0;

if (remainingSeconds > HOUR_INT) {
  timeUnit = HOUR_INT;
  suffix = " hours remaining";
} else if (remainingSeconds > MINUTE_INT) {
  timeUnit = MINUTE_INT;
  suffix = " minutes remaining";
} else if (remainingSeconds > SECOND_INT) {
  timeUnit = SECOND_INT;
  suffix = " seconds remaining";
}

return this.divideAndRound(remainingSeconds, timeUnit, prefixSeparator, suffix);
```
it would also require adding two more arguments to divideAndRound:

```javascript
divideAndRound(unit, divide, prefix = '', suffix = '') {
  const num = Math.round(unit / divide);

  return `${prefix}${num}${suffix}`;
}
```

Reply:
Yeah, I could've done a better job there, my mistake.

Be aware that the current base case of retuning an empty string is actually used. I would also prefer not to change the divideAndRound definition since it is used elsewhere. I've pushed an update that cleans it up a little bit and makes everything explicit.

Reply:
👍